keps/sig-instrumentation/4200-healthcheck-metrics/README.md

```diff
@@ -0,0 +1,297 @@
+health check data in prometheus format.
+
+
+### Risks and Mitigations
+
+This is a separate endpoint, it does not have to be used. The risk of adding metrics is generally cardinality, but in this
+case we are proposing known dimensions to the metrics, specifically:
+
+1. status - one of `Success`, `Error`, `Pending`
+1. type - one of `livez`, `readyz`, `healthz`
+1. name - the known name of the health check. AFAIK, these are all static strings in the Kubernetes codebase, therefore bounded in cardinality
+
+
+## Design Details
+
+
+### Test Plan
+
+
+##### Unit tests
+
+- [ ] ensure that healthcheck states are reset for gauges on write
+- [ ] ensure that counters properly retain state of all seen healthchecks
+
+##### Integration tests
+
+- [ ] ensure existence of healthcheck endpoint
+
+##### e2e tests
+
+
+- [ ] ensure existence of healthcheck endpoint
+
+
+### Graduation Criteria
+
+
+## Production Readiness Review Questionnaire
+
+
+### Feature Enablement and Rollback
+
+We can feature gate the api and the registry for this metrics endpoint, such that it is disabled by default and toggleable only by feature-gate. Given the lazy
+nature of the Kubernetes metric framework, this will ensure that we do not encounter any memory leaks.
+
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+Via feature gate.
+
+
+###### Does enabling the feature change any default behavior?
+
+Yes, it exposes an endpoint which can then be scraped.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, but it will remove the endpoint.
+
+### Monitoring Requirements
+
+I am proposing a metric/series of metrics here.
+
+### Scalability
+
+We already hit the readyz/healthz/livez endpoints of our control-plane components frequently, this KEP only 
+adds instrumentation of these endpoints' results.
+
+###### Will enabling / using this feature result in any new API calls?
+
+Yes, we are proposing that this health metrics are surfaced in each component under `/health/metrics` which 
+will have to be consumed for the feature to be useful. However, this should be relatively innocuous since 
+it will an isolated endpoint strictly for the purpose of surfacing health metrics.
+
+###### Will enabling / using this feature result in introducing new API types?
+
+No.
+
+###### Will enabling / using this feature result in any new calls to the cloud provider?
+
+No. 
+
+###### Will enabling / using this feature result in increasing size or count of the existing API objects?
+
+No. 
+
+###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?
+
+No. This should, in theory, reduce calls to health endpoints since SLIs need to be calculated currently by directly hitting
+and parsing out the results of our existing health check endpoints, which adds to the total number of calls (since kubelet 
+also hits these endpoints). 
```

Comment:
And presumably, any issue in threading caused by multiple callers to the endpoints would already have been discovered.

Reply:
Will there be novel plumbing in components between metrics and healthz that would differ in behavior from the current call pattern that we have to consider?

Reply:
We could, in theory, but I was mostly planning to just wire the metrics into the healthz/livez/readyz endpoints. The granularity of kubelet scraping interval (which causes the writes to the metric) should be sufficiently granular for reliable SLI data.